backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/types/dsl/impl/ElasticsearchBigIntegerIndexFieldTypeContext.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.backend.elasticsearch.types.dsl.impl;
+
+import java.math.BigInteger;
+
+import org.hibernate.search.backend.elasticsearch.document.model.impl.esnative.DataType;
+import org.hibernate.search.backend.elasticsearch.document.model.impl.esnative.PropertyMapping;
+import org.hibernate.search.backend.elasticsearch.types.codec.impl.ElasticsearchBigIntegerFieldCodec;
+import org.hibernate.search.backend.elasticsearch.types.impl.ElasticsearchIndexFieldType;
+import org.hibernate.search.backend.elasticsearch.types.predicate.impl.ElasticsearchStandardFieldPredicateBuilderFactory;
+import org.hibernate.search.backend.elasticsearch.types.projection.impl.ElasticsearchStandardFieldProjectionBuilderFactory;
+import org.hibernate.search.backend.elasticsearch.types.sort.impl.ElasticsearchStandardFieldSortBuilderFactory;
+import org.hibernate.search.engine.backend.types.converter.FromDocumentFieldValueConverter;
+import org.hibernate.search.engine.backend.types.converter.ToDocumentFieldValueConverter;
+
+class ElasticsearchBigIntegerIndexFieldTypeContext
+		extends AbstractElasticsearchScalarFieldTypeContext<ElasticsearchBigIntegerIndexFieldTypeContext, BigInteger> {
+
+	ElasticsearchBigIntegerIndexFieldTypeContext(ElasticsearchIndexFieldTypeBuildContext buildContext) {
+		super( buildContext, BigInteger.class, DataType.LONG );
```

Comment:
I understand there is no support for arbitrary-precision integers in Elasticsearch, but if we want to make it work regardless, we should try to find a compromise.

Indexing as long is not a compromise, it's straight up surrender: users need to rely on `BigInteger` when they need values that *cannot* fit into a `long`. Which means indexing, here, will always fail.

For `BigDecimal`, the compromise was that we provided higher precision than `double` (64 bits instead of 53 bits), but still limited precision, and with a scale that is as limited as in double (in the case of the Elasticsearch backend) or to what an integer can hold (in the case of the Lucene backend).

For `BigInteger`, we have several possible approaches:

1. Index as a string or byte[], get arbitrary precision for projections at the cost of meaningless range queries or sorts.
2. Index as a `scaled_float`, get arbitrary precisions for projections (I think?) at the cost of lower precision for range queries or sorts.

Solution number 2 seems better than nothing. It wouldn't provide arbitrary precision, but at least it would address *some* use cases. Which is better than indexing as a long.

For Lucene, I think we could theoretically ask users to give us a precision upfront, and then index as an array of integers. I'm not sure how well this would work, but it would be an option. We should probably start with a long and an integer scale for now, though (just like `BigDecimal`). Later we will be able to add an explicit "precision" parameter and to use a different encoding when that precision cannot be achieved with a `long`.

Reply:
@yrodiere would option 2 consist in using always a decimal scale less or equal to 0?

Reply:
Yes. I guess we could allow strictly positive scales if it's easier that way, but that wouldn't be very useful.

Reply:
Keeping scales negative or at most equals to zero, it would be more consistent with use of scale for BigDecimal. The only differences that I would make are: 

1 To forbid positive scales for BigInteger, since decimal digits are not supported with them.  
2. To make `0` the default value for scale, since I think most of the time they are used as long.

Reply:
> To make `0` the default value for scale, since I think most of the time they are used as long.

I don't know where you're getting this from? The purpose of a `BigInteger` is to hold values that cannot be held by a `long`, be it for storage, or just for intermediary values in calculations. Therefore it's unlikely that a `BigInteger` used in an application will be convertible to a long without any scale: it will be greater than `Long.MAX_VALUE`.

Please don't use `0` as a default. If you're right we can add it later if we get new information, but if I'm right we won't be able to remove the default easily and many users will use a wrong mapping an only notice the problem at runtime when indexing.